inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/PulsarSource.java

```diff
@@ -0,0 +1,312 @@
+            0, Integer.MAX_VALUE,
+            1L, TimeUnit.SECONDS,
+            new SynchronousQueue<>(),
+            new AgentThreadFactory("pulsar-source"));
+    private BlockingQueue<SourceData> queue;
+    public InstanceProfile profile;
+    private int maxPackSize;
+    private String taskId;
+    private String instanceId;
+    private String topic;
+    private String serviceUrl;
+    private String scanStartupMode;
+    private PulsarClient pulsarClient;
+    private Long timestamp;
+    Map<Integer, Long> partitionOffsets = new HashMap<>();
+    private volatile boolean running = false;
+    private volatile boolean runnable = true;
+    private volatile AtomicLong emptyCount = new AtomicLong(0);
+
+    private final Integer CACHE_QUEUE_SIZE = 100000;
+    private final Integer READ_WAIT_TIMEOUT_MS = 10;
+    private final Integer EMPTY_CHECK_COUNT_AT_LEAST = 5 * 60;
+    private final Integer BATCH_TOTAL_LEN = 1024 * 1024;
+
+    private final Integer CORE_THREAD_PRINT_INTERVAL_MS = 1000;
+    private boolean isRealTime = false;
+    private boolean isRestoreFromDB = false;
+
+    public PulsarSource() {
+    }
+
+    @Override
+    public void init(InstanceProfile profile) {
+        try {
+            LOGGER.info("PulsarSource init: {}", profile.toJsonStr());
+            this.profile = profile;
+            super.init(profile);
+            String cycleUnit = profile.get(TASK_CYCLE_UNIT);
+            if (cycleUnit.compareToIgnoreCase(CycleUnitType.REAL_TIME) == 0) {
+                isRealTime = true;
+                cycleUnit = CycleUnitType.HOUR;
+            }
+            queue = new LinkedBlockingQueue<>(CACHE_QUEUE_SIZE);
+            maxPackSize = profile.getInt(PROXY_PACKAGE_MAX_SIZE, DEFAULT_PROXY_PACKAGE_MAX_SIZE);
+            taskId = profile.getTaskId();
+            instanceId = profile.getInstanceId();
+            topic = profile.getInstanceId();
+            serviceUrl = profile.get(TASK_PULSAR_SERVICE_URL);
+            scanStartupMode = profile.get(TASK_PULSAR_SCAN_STARTUP_MODE);
+            timestamp = profile.getLong(TASK_PULSAR_RESET_TIME, 0);
+            pulsarClient = PulsarClient.builder().serviceUrl(serviceUrl).build();
+
+            EXECUTOR_SERVICE.execute(run());
+        } catch (Exception ex) {
+            stopRunning();
+            throw new FileException("error init stream for " + topic, ex);
+        }
+    }
+
+    private Runnable run() {
+        return () -> {
+            AgentThreadFactory.nameThread("pulsar-source-" + taskId + "-" + instanceId);
+            running = true;
+            try {
+                try (Consumer<byte[]> consumer = pulsarClient.newConsumer(Schema.BYTES)
+                        .topic(topic)
+                        .subscriptionName("inlong-agent-" + taskId)
+                        .subscriptionInitialPosition(SubscriptionInitialPosition.valueOf(scanStartupMode))
+                        .subscriptionType(SubscriptionType.Exclusive)
```

Comment:
Exclusive subscribe mode limits the consumer count must =1 under a subscription to a topic, Please check if we can meet this requirement now.

Reply:
Yes, maybe Shared subscribe mode is more suitable.